### PR TITLE
Add tracing and jaeger

### PR DESCRIPTION
**Pull request by sevenzing**
jaeger usage examples:

image

### CODE REVIEW

File: verification/src/tracer.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        tracing_subscriber::EnvFilter::builder()
+            .with_default_directive(LevelFilter::INFO.into())
+            .from_env_lossy(),
+    );
+    let tracer = init_jaeger_tracer().expect("failed to init tracer");
+    tracing_subscriber::registry()
+        // output logs (tracing) to stdout with log level taken from env (default is INFO)
+        .with(stdout)
+        // output traces to jaeger with default log level (default is TRACE)
+        .with(tracing_opentelemetry::layer().with_tracer(tracer))
```
Comment (leviska) [inline verification/src/tracer.rs:15]: may be add tracing endpoint too? like with metrics

File: verification/src/tracer.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+use opentelemetry::{sdk::trace::Tracer, trace::TraceError};
+use tracing_subscriber::{filter::LevelFilter, layer::SubscriberExt, prelude::*};
+
+pub fn init_logs() {
```
Comment (leviska) [inline verification/src/tracer.rs:4]: May be add "enabled" option, so people could disable tracing altogether (or even disable them by default)
but leave the logs

Comment (sevenzing) [reply verification/src/tracer.rs:4]: what do you mean by `disable tracing`? we use tracing for logs. you mean disable jaeger?

File: verification/src/tracer.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            // output traces to jaeger with default log level (default is DEBUG)
+            .with(
+                tracing_opentelemetry::layer()
+                    .with_tracer(tracer)
+                    .with_filter(LevelFilter::DEBUG),
+            )
+            .try_init()
+    } else {
+        registry.try_init()
+    }
```
Comment (sevenzing) [inline verification/src/tracer.rs:29]: don't like this way, but Idk how to make it without else

File: verification/src/config.rs [hunk 4]
```diff
@@ -87,6 +88,22 @@ impl Default for MetricsConfiguration {
     }
 }
 
+#[derive(Deserialize, Clone)]
+#[serde(default)]
+pub struct TracingConfiguration {
```
Comment (sevenzing) [inline verification/src/config.rs:93]: should I rename it to JaegerConfiguration? not sure about it

File: verification/src/tracer.rs (hunk 3 above)
Comment (leviska) [reply verification/src/tracer.rs:29]: doesn't this work
```rust
let registry = if config.enable_jaeger {
  let tracer = ...;
  registry.with(...)
} else {
  registry
}
registry.try_init().expect(..);
```

**Review by leviska — APPROVED** [review]

File: verification/src/tracer.rs (hunk 3 above)
Comment (sevenzing) [reply verification/src/tracer.rs:29]: Of course I tried this, but problem is that `.with()` takes self and returns not Self but `SomeStructure<Self>`, and you cannot case `registry` to `registry.with(...)`

Comment (leviska) [reply verification/src/tracer.rs:29]: well it seems you can only do this using dyn (if you even can), which in your case is actually worse imo
but I'll provide it jfyi
something like
```rust
let registry: Box<dyn SubscriberInitExt> = Box::new(
  if ... {
    let tracer = ...;
    registry.with(..) as dyn SubscriberInitExt
  } else {
    registry as dyn SubscriberInitExt
  }
);
```
or something like that

File: verification/src/compiler/compilers.rs [hunk 5]
```diff
@@ -31,19 +32,31 @@ impl Compilers {
...
         &self,
         compiler_version: &compiler::Version,
         input: &CompilerInput,
     ) -> Result<CompilerOutput, Error> {
-        let solc_path = match self.cache.get(&*self.fetcher, compiler_version).await {
+        let solc_path_result = {
+            let span = tracing::debug_span!("get solc_path from cache");
+            self.cache
+                .get(self.fetcher.as_ref(), compiler_version)
+                .instrument(span)
```
Comment (leviska) [inline verification/src/compiler/compilers.rs:45]: For the most cases imo it's better to put the span inside the function, if possible
Like here, just put it inside cache

File: verification/src/compiler/download_cache.rs [hunk 6]
```diff
@@ -47,7 +49,8 @@ impl DownloadCache {
             }
             None => {
                 let _timer = metrics::COMPILER_FETCH_TIME.start_timer();
-                self.fetch(fetcher, ver).await
+                let span = tracing::debug_span!("fetch compiler", ver = ver.to_string());
+                self.fetch(fetcher, ver).instrument(span).await
```
Comment (leviska) [inline verification/src/compiler/download_cache.rs:53]: it's fine as of now, but just a question
I saw there is a `#instrument` macro, which will automatically add the span. Is there something like it for the trait function? Not the impl block, but the trait, like
```rust
trait Foo {
  #[instrument]
  fn foo();
}
```
If it exists, then I suggest changing the span to such macro, If not, current impl is okay

File: verification/src/compiler/fetcher.rs [hunk 7]
```diff
@@ -44,15 +45,12 @@ fn create_executable(path: &Path) -> Result<File, std::io::Error> {
...
 pub fn validate_checksum(bytes: &Bytes, expected: H256) -> Result<(), Mismatch<H256>> {
-    let start = std::time::Instant::now();
-
-    let found = Sha256::digest(bytes);
-    let found = H256::from_slice(&found);
-
-    let took = std::time::Instant::now() - start;
-    // TODO: change to tracing
-    log::debug!("check hashsum of {} bytes took {:?}", bytes.len(), took,);
+    let found = {
```
Comment (leviska) [inline verification/src/compiler/fetcher.rs:50]: why change this to a code block?

File: verification/src/compiler/list_fetcher.rs [hunk 8]
```diff
@@ -105,6 +106,7 @@ impl ListFetcher {
         Ok(Self { versions, folder })
     }
 
+    #[instrument(skip(self) level = "debug")]
```
Comment (leviska) [inline verification/src/compiler/list_fetcher.rs:109]: `    #[instrument(skip(self), level="debug")]` (comma)?

File: verification/src/compiler/s3_fetcher.rs [hunk 9]
```diff
@@ -97,6 +98,7 @@ impl S3Fetcher {
         })
     }
 
+    #[instrument(skip(self) level = "debug")]
```
Comment (leviska) [inline verification/src/compiler/s3_fetcher.rs:101]: comma?

File: verification/src/compiler/download_cache.rs (hunk 6 above)
Comment (sevenzing) [reply verification/src/compiler/download_cache.rs:53]: no, I think it's not possible to add instrument to trait function

File: verification/src/compiler/versions_fetcher.rs [hunk 10]
```diff
@@ -84,7 +84,7 @@ impl<T: PartialEq + Send + Sync + 'static> VersionsRefresher<T> {
         fetcher: Arc<F>,
         cron_schedule: Schedule,
     ) {
-        log::info!("spawn version refresh job");
+        tracing::info!("spawn version refresh job");
         scheduler::spawn_job(cron_schedule, "refresh compiler versions", move || {
             let fetcher = fetcher.clone();
```
Comment (leviska) [inline verification/src/compiler/versions_fetcher.rs:89]: can we create new trace here? so we could search for scheduled things later in jaeger

File: verification/src/compiler/compilers.rs (hunk 5 above)
Comment (sevenzing) [reply verification/src/compiler/compilers.rs:45]: this span just doesn't attach to current trace without it. here is more about it:
https://docs.rs/tracing/latest/tracing/struct.Span.html#in-asynchronous-code

File: verification/src/compiler/versions_fetcher.rs [hunk 11]
```diff
@@ -32,7 +32,7 @@ impl<T> VersionsRefresher<T> {
 
 impl<T: PartialEq> VersionsRefresher<T> {
     pub async fn refresh<F: VersionsFetcher<Versions = T>>(&self, fetcher: &F) {
-        log::info!("looking for new compilers versions");
+        tracing::info!("looking for new compilers versions");
```
Comment (leviska) [inline verification/src/compiler/versions_fetcher.rs:35]: add span here may be

File: verification/src/compiler/versions_fetcher.rs [hunk 12]
```diff
@@ -32,7 +32,7 @@ impl<T> VersionsRefresher<T> {
...
         let refresh_result = fetcher.fetch_versions().await;
```
Comment (leviska) [inline verification/src/compiler/versions_fetcher.rs:36]: add span for this await

File: verification/src/http_server/handlers/verification/solidity/contract_verifier.rs [hunk 13]
```diff
@@ -68,13 +70,15 @@ pub(crate) async fn compile_and_verify_handler(
...
 
+#[instrument(skip(compilers, verifier, input), level = "debug")]
 async fn compile_and_verify(
     compilers: &Compilers,
     verifier: &Verifier,
     input: &Input<'_>,
 ) -> Result<VerificationSuccess, CompileAndVerifyError> {
     let compiler_output = compilers
         .compile(&input.compiler_version, &input.compiler_input)
+        .in_current_span()
```
Comment (leviska) [inline verification/src/http_server/handlers/verification/solidity/contract_verifier.rs:81]: is this doing anything? like, you've already got a span for whole function, and this call isn't a separate span, what happens if you add this line? O.o

File: verification/src/compiler/compilers.rs (hunk 5 above)
Comment (leviska) [reply verification/src/compiler/compilers.rs:45]: I'm talking about just writing the macro for this function
image

File: verification/src/compiler/versions_fetcher.rs (hunk 10 above)
Comment (leviska) [reply verification/src/compiler/versions_fetcher.rs:89]: (not the span, but a trace)

File: verification/src/compiler/fetcher.rs [hunk 14]
```diff
@@ -72,21 +70,26 @@ pub async fn write_executable(
...
-                } else {
-                    Err(e)
-                }
-            })?;
-            let mut file = create_executable(file.as_path())?;
-            std::io::copy(&mut data.as_ref(), &mut file)?;
-            Ok(())
+        let span = tracing::debug_span!("save executable");
+        tokio::task::spawn_blocking(move || {
+            span.in_scope(|| {
```
Comment (leviska) [inline verification/src/compiler/fetcher.rs:75]: I know you've talked about this, but what's the difference between `in_scope` and `enter`?
I mean, in the code it's just the same
https://docs.rs/tracing/latest/src/tracing/span.rs.html#785-788
https://docs.rs/tracing/latest/src/tracing/span.rs.html#1100-1103

File: verification/src/compiler/versions_fetcher.rs (hunk 11 above)
Comment (leviska) [reply verification/src/compiler/versions_fetcher.rs:35]: I mean for the whole function, `#instrument` may be better

File: verification/example_config.toml [hunk 15]
```diff
@@ -13,5 +13,10 @@ refresh_versions_schedule = "0 0 * * * * *"
 compilers_list_url = "https://solc-bin.ethereum.org/linux-amd64/list.json"
 
 [metrics]
+enabled = false
 endpoint = "/metrics"
 addr = "0.0.0.0:6060"
+
+[tracing]
+enable_jaeger = false
+jaeger_agent = "localhost:6831"
```
Comment (rimrakhimov) [inline verification/example_config.toml:22]: I suggest to name the section `jaeger` and use `enabled` and `agent_addr` as keys.

As another option we may use `tracing.jaeger` as a subsection, but not sure whether it make sense now, as we do not have any other options for `tracing` section

**Review by rimrakhimov — APPROVED** [review]

File: verification/src/compiler/fetcher.rs (hunk 14 above)
Comment (sevenzing) [reply verification/src/compiler/fetcher.rs]: changed to `.enter()`

**Review by leviska — APPROVED** [review]

**Pull request merged**